### ISSUE TITLE
+= for variables

### ISSUE THREAD

**Issue by elyisgreat** [alda-lang/alda#248]
take the following code:

```
notes = c d
notes = e f
piano: notes
```

This will play just the latter two notes, however what if one wanted to play all of them? I suggest a += operator to allow for appending to variables. This allows for the versatility of variables as well as adding more flexibility to writing style. An alternative that one could use is 

```
piano: c d
piano: e f
```

this way, one could group the variable parts together:

```
notes = c d
notes += e f
piano: notes
```

An example with a longer score:

```
(tempo! 180)

piano: (vol 60)
piano "melodicPiano": (vol 90)
violin: (vol 90)
trumpet: (vol 90)
flute: (vol 80)
percussion: (vol 30)

piano: [o5 e4/g/b g/b/>d< g/b/>e< b/>d/f+< g/b/>d< g/b/>e< b/>d/f+< e/g/b g/b/>d< b/>d/f+< g/b/>e< g/b/>d<]*2
melodicPiano: [r2.]*4 o4 <e>/e <b>/b <a>/a d/>d<

piano: o3 g2./b/>e< f+/b/>e-< g/b/d e/a/>c+< e/a/>c< f+/a/>d< d/g/b e-/f+/b e/g/b e/g/b e/a/>c< e/g/b e/a/>c< a/>c/e< f+/b/>e-< f+/b/>e-<
violin: [r2.]*8 o5 b b b4 a g a2. g4 f+ e d f+ d < b1. > b2. b4 > e f+ g2 f+4 e2. c2 < b4 a g a b1.

percussion: [r2.]*24 o2 [d4]*45 r2.
violin: (vol 60) (quant 50) o3 [a4/>c/e<]*6 [a4/>d/f+<]*6 [b4/>d/f+<]*6 [g4/b/>e<]*6 [a4/>c/e<]*6 [b4/>e-/f+<]*6 (quant 90) b2./>e/g d/f+/a e/a-/b r (vol 90)
trumpet: [r2.]*24 o4 c2.~2 g4 f+2.~2 e4 f+2. d2 < b4 > d1. e2. c4 d e e-2. e-4 e f+ a2. g2 a4 b2. r
flute: [r2.]*30 o5 r4 g8/b a/>c< b4/>d< b/>d< a/>c< g/b [r2.]*8

melodicPiano: [r2.]*31 o5 a-4 a b > c2. c4 < b a g2 b4 d2. g g4 f+ e d2 f+4 < b2. > c4 < b > c e c > c < b2. e-4 b a g f+ g a g a b1.
piano: [r2.]*15 o4 e4 f+ a- a > c e g f+ e d2. < b c4 e g > c < b a g2. d a4 e a > c < a > c e-2./f+ < f+/b e/g f+/a e-1./f+/b
```

With += you could write it like so, preserving the original structure:

```
(tempo! 180)

pianoPart = (vol 60)
melodicPianoPart = (vol 90)
violinPart = (vol 90)
trumpetPart = (vol 90)
flutePart = (vol 80)
percussionPart = (vol 30)

pianoPart += [o5 e4/g/b g/b/>d< g/b/>e< b/>d/f+< g/b/>d< g/b/>e< b/>d/f+< e/g/b g/b/>d< b/>d/f+< g/b/>e< g/b/>d<]*2
melodicPianoPart += [r2.]*4 o4 <e>/e <b>/b <a>/a d/>d<

pianoPart += o3 g2./b/>e< f+/b/>e-< g/b/d e/a/>c+< e/a/>c< f+/a/>d< d/g/b e-/f+/b e/g/b e/g/b e/a/>c< e/g/b e/a/>c< a/>c/e< f+/b/>e-< f+/b/>e-<
violinPart += [r2.]*8 o5 b b b4 a g a2. g4 f+ e d f+ d < b1. > b2. b4 > e f+ g2 f+4 e2. c2 < b4 a g a b1.

percussionPart += [r2.]*24 o2 [d4]*45 r2.
violinPart += (vol 60) (quant 50) o3 [a4/>c/e<]*6 [a4/>d/f+<]*6 [b4/>d/f+<]*6 [g4/b/>e<]*6 [a4/>c/e<]*6 [b4/>e-/f+<]*6 (quant 90) b2./>e/g d/f+/a e/a-/b r (vol 90)
trumpetPart += [r2.]*24 o4 c2.~2 g4 f+2.~2 e4 f+2. d2 < b4 > d1. e2. c4 d e e-2. e-4 e f+ a2. g2 a4 b2. r
flutePart += [r2.]*30 o5 r4 g8/b a/>c< b4/>d< b/>d< a/>c< g/b [r2.]*8

melodicPianoPart += [r2.]*31 o5 a-4 a b > c2. c4 < b a g2 b4 d2. g g4 f+ e d2 f+4 < b2. > c4 < b > c e c > c < b2. e-4 b a g f+ g a g a b1.
pianoPart += [r2.]*15 o4 e4 f+ a- a > c e g f+ e d2. < b c4 e g > c < b a g2. d a4 e a > c < a > c e-2./f+ < f+/b e/g f+/a e-1./f+/b

piano: pianoPart
piano "melodicPiano": melodicPianoPart
violin: violinPart
trumpet: trumpetPart
flute: flutePart
percussion: percussionPart
```

However with the current variable syntax, it would have to be written like this:

```
(tempo! 180)

pianoPart = [(vol 60) [o5 e4/g/b g/b/>d< g/b/>e< b/>d/f+< g/b/>d< g/b/>e< b/>d/f+< e/g/b g/b/>d< b/>d/f+< g/b/>e< g/b/>d<]*2

o3 g2./b/>e< f+/b/>e-< g/b/d e/a/>c+< e/a/>c< f+/a/>d< d/g/b e-/f+/b e/g/b e/g/b e/a/>c< e/g/b e/a/>c< a/>c/e< f+/b/>e-< f+/b/>e-<

[r2.]*15 o4 e4 f+ a- a > c e g f+ e d2. < b c4 e g > c < b a g2. d a4 e a > c < a > c e-2./f+ < f+/b e/g f+/a e-1./f+/b]

melodicPianoPart = [(vol 90) [r2.]*4 o4 <e>/e <b>/b <a>/a d/>d<

[r2.]*31 o5 a-4 a b > c2. c4 < b a g2 b4 d2. g g4 f+ e d2 f+4 < b2. > c4 < b > c e c > c < b2. e-4 b a g f+ g a g a b1.]

violinPart = [(vol 90) [r2.]*8 o5 b b b4 a g a2. g4 f+ e d f+ d < b1. > b2. b4 > e f+ g2 f+4 e2. c2 < b4 a g a b1.

(vol 60) (quant 50) o3 [a4/>c/e<]*6 [a4/>d/f+<]*6 [b4/>d/f+<]*6 [g4/b/>e<]*6 [a4/>c/e<]*6 [b4/>e-/f+<]*6 (quant 90) b2./>e/g d/f+/a e/a-/b r (vol 90)]

trumpetPart = (vol 90) [r2.]*24 o4 c2.~2 g4 f+2.~2 e4 f+2. d2 < b4 > d1. e2. c4 d e e-2. e-4 e f+ a2. g2 a4 b2. r

flutePart = (vol 80) [r2.]*30 o5 r4 g8/b a/>c< b4/>d< b/>d< a/>c< g/b [r2.]*8

percussionPart = (vol 30) [r2.]*24 o2 [d4]*45 r2.


piano: pianoPart
piano "melodicPiano": melodicPianoPart
violin: violinPart
trumpet: trumpetPart
flute: flutePart
percussion: percussionPart
```

**Comment by jimcheetham** [comment]
I would have thought that having the variable as a reference on the right-hand side would have been just as useful as a dedicated concat operator :+1: 

notes = a b > c
notes = notes d e f
piano: notes

However on my v1.0.0.-rc23 that breaks ...

```
                  clojure.core/reduce      core.clj: 6544
                                  ...                    
alda.lisp.events.variable/eval5077/fn  variable.clj:   10
                                  ...                    
alda.lisp.events.sequence/eval5057/fn  sequence.clj:   13
```
these three lines repeated about 170 times and then

```
                  clojure.core/reduce      core.clj: 6544
                                  ...                    
alda.lisp.events.variable/eval5077/fn  variable.clj:   10
                                  ...                    
alda.lisp.events.sequence/eval5057/fn  sequence.clj:   13
                  clojure.core/reduce      core.clj: 6544
                                  ...                    
    alda.lisp.model.event/eval2745/fn     event.clj:   21
java.lang.StackOverflowError: 
```

... so I guess not :-)

As far as syntax is concerned, x += y is usually sugar for x = x + y, so fixing x = x y first seems useful

**Comment by daveyarwood** [comment]
> I would have thought that having the variable as a reference on the right-hand side would have been just as useful as a dedicated concat operator :+1: 

That was my first thought. Good to know that that's broken! :P  

I think I know what the problem is... I'll see if I can fix it. I think `x = x y` is a good solution for this -- would prefer not to add additional syntax, if we can avoid it.

**Comment by elyisgreat** [comment]
Eh; `x += y` looks nicer. I'd add support for both, but `x = x y` does do the job.

**Comment by jimcheetham** [comment]
One advantage of x = x y is that when generalised properly, you get x = w x
y as well, meaning you get a more flexible way to use the variables in
assignment than just the append implied by +=.

piano:
hook = a b >c<
verse = a/b a/b hook
verse
hook = >c< b a
verse

(this currently works in -rc23 and the second playback of 'verse' sounds
different to the first, as you might expect)

Redefining hook in this case means that verse plays differently, too -
obviously a run-time expansion of the variable, rather than a declare-time
one. This is probably the cause of the break when "notes = notes a b c"
thing I found, the run-time expansion of 'notes' included a reference to
itself.

However, fixing that to a declare-time situation means that we'd be better
off not calling these things variables at all, they'd effectively be
immutable values, and probably should be called 'references' or
'declarations'. It depends on how much of a programmer you want an alda
user to be :-)


-jim

**Comment by daveyarwood** [comment]
> However, fixing that to a declare-time situation means that we'd be better
off not calling these things variables at all, they'd effectively be
immutable values, and probably should be called 'references' or
'declarations'.

It's funny, I've had that same thought. As a proponent of functional programming, my inclination is to always work with immutable values, i.e. after assigning a name to a value, that value never changes, and is always associated with that name. This is how I would intend Alda "variables" to be used (at least, it's how I would use them), but at the same time, I'm not particularly against letting people change the value after the initial assignment. My view on this is, different people will have different ideas about how to use variables in Alda, so it would probably be best to be flexible and allow treating them as immutable values (so not really a "variable," but more of an "alias" or a "definition," I guess), or as mutable values (true variables).

What we have right now in 1.0.0-rc23 is, purely by accident (I never thought to try something like `x = x y`), variables that work like pointers. I hate working with pointers, as they are often a source of confusion. Redefining a variable after its initial definition can alter the value of any variables which depend on it, which is horrible. Instead, I'd like to get the kind of behavior that you get with most modern programming languages -- when you define a value, you can use other named values (whether they're called variables or constants or whatever) in the definition, and you can be confident that your defined value will never change unless you explicitly redefine it.

To use Python or Ruby (either will work) as an example:

```
x = 42
y = x + 1 
# x is 42, y is 43

x = 90
# x is now 90, but y is still 43 because we didn't redefine y
```

I find this type of logic way more intuitive, and I would guess that most other programmers (even beginners) would expect it to work this way.

**Comment by elyisgreat** [comment]
Agreed. Immutable variable logic makes the most sense here. 

I would imagine that replacing `x += y` with `x = x y` internally wouldn't be difficult. Although you can let me know otherwise.

**Comment by daveyarwood** [comment]
The variable scoping issues are fixed in [1.0.0-rc24](https://github.com/alda-lang/alda/releases/tag/1.0.0-rc24).

I decided not to add a `+=` operator, since I'm not convinced that it's something the language needs. I have no doubt that it would be easy to implement, but it's a question of design. I'll admit that I'm totally playing my "opinionated language creator/[BDFL](https://en.wikipedia.org/wiki/Benevolent_dictator_for_life)" card here, but I think that using variables in a way where you define them and then redefine them is an antipattern. I think what we have in 1.0.0-rc24, where it's at least possible to do `foo = foo bar`, is a good compromise for people who really want something that behaves like `+=`, but I'd prefer to steer people away from doing this because I think in most/all cases, there are better ways to write your score.

@elyisgreat if you want, you can email me or ping me on Slack if you need any help with figuring out a better way to write scores where you would otherwise find yourself wanting to use a `+=` operator -- I'm confident that what we have currently should give score writers a great deal of flexibility of expression, it's just a matter of knowing how to use them. I think the bigger issue might be documenting some "best practices" about using variables.

**Comment by elyisgreat** [comment]
Actually the reason I wrote the score with this `+=` syntax was in order to make it easier to repeat. See this discussion we had in #202 , before variables were implemented.
